formatprocessor/formatprocessor.go

```diff
@@ -0,0 +1,532 @@
+		h264.NALUTypeSliceExtension,
+		h264.NALUTypeSliceExtensionDepth,
+		h264.NALUTypeReserved22,
+		h264.NALUTypeReserved23,
+		h264.NALUTypeSTAPB,
+		h264.NALUTypeMTAP16,
+		h264.NALUTypeMTAP24,
+		h264.NALUTypeFUA,
+		h264.NALUTypeFUB:
+		fallthrough
+
+	default:
+		return nil, nil
+	}
+}
+
+type formatProcessorH264 struct {
+	udpMaxPayloadSize int
+	format            *format.H264
+
+	encoder *rtph264.Encoder
+	decoder *rtph264.Decoder
+}
+
+func newH264(
+	udpMaxPayloadSize int,
+	forma *format.H264,
+	generateRTPPackets bool,
+) (*formatProcessorH264, error) {
+	t := &formatProcessorH264{
+		udpMaxPayloadSize: udpMaxPayloadSize,
+		format:            forma,
+	}
+
+	if generateRTPPackets {
+		err := t.createEncoder(nil, nil)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return t, nil
+}
+
+func (t *formatProcessorH264) createEncoder(
+	ssrc *uint32,
+	initialSequenceNumber *uint16,
+) error {
+	t.encoder = &rtph264.Encoder{
+		PayloadMaxSize:        t.udpMaxPayloadSize - 12,
+		PayloadType:           t.format.PayloadTyp,
+		SSRC:                  ssrc,
+		InitialSequenceNumber: initialSequenceNumber,
+		PacketizationMode:     t.format.PacketizationMode,
+	}
+	return t.encoder.Init()
+}
+
+func (t *formatProcessorH264) updateTrackParametersFromRTPPacket(payload []byte) {
+	sps, pps := rtpH264ExtractParams(payload)
+
+	if (sps != nil && !bytes.Equal(sps, t.format.SPS)) ||
+		(pps != nil && !bytes.Equal(pps, t.format.PPS)) {
+		if sps == nil {
+			sps = t.format.SPS
+		}
+		if pps == nil {
+			pps = t.format.PPS
+		}
+		t.format.SafeSetParams(sps, pps)
+	}
+}
+
+func (t *formatProcessorH264) updateTrackParametersFromAU(au [][]byte) {
+	sps := t.format.SPS
+	pps := t.format.PPS
+	update := false
+
+	for _, nalu := range au {
+		typ := h264.NALUType(nalu[0] & 0x1F)
+
+		switch typ {
+		case h264.NALUTypeSPS:
+			if !bytes.Equal(nalu, sps) {
+				sps = nalu
+				update = true
+			}
+
+		case h264.NALUTypePPS:
+			if !bytes.Equal(nalu, pps) {
+				pps = nalu
+				update = true
+			}
+
+		case h264.NALUTypeNonIDR,
+			h264.NALUTypeDataPartitionA,
+			h264.NALUTypeDataPartitionB,
+			h264.NALUTypeDataPartitionC,
+			h264.NALUTypeIDR,
+			h264.NALUTypeSEI,
+			h264.NALUTypeAccessUnitDelimiter,
+			h264.NALUTypeEndOfSequence,
+			h264.NALUTypeEndOfStream,
+			h264.NALUTypeFillerData,
+			h264.NALUTypeSPSExtension,
+			h264.NALUTypePrefix,
+			h264.NALUTypeSubsetSPS,
+			h264.NALUTypeReserved16,
+			h264.NALUTypeReserved17,
+			h264.NALUTypeReserved18,
+			h264.NALUTypeSliceLayerWithoutPartitioning,
+			h264.NALUTypeSliceExtension,
+			h264.NALUTypeSliceExtensionDepth,
+			h264.NALUTypeReserved22,
+			h264.NALUTypeReserved23,
+			h264.NALUTypeSTAPB,
+			h264.NALUTypeMTAP16,
+			h264.NALUTypeMTAP24,
+			h264.NALUTypeFUA,
+			h264.NALUTypeSTAPA,
+			h264.NALUTypeFUB:
+		default:
+		}
+	}
+
+	if update {
+		t.format.SafeSetParams(sps, pps)
+	}
+}
+
+func (t *formatProcessorH264) remuxAccessUnit(au [][]byte) [][]byte {
```

Comment:
[nit] Is it possible to have single pass over the au here? Also maybe a simpler way to pack bytes rather than index handling?

Something like this:
```
func (t *formatProcessorH264) remuxAccessUnit(au [][]byte) [][]byte {
    var filteredNALUs [][]byte
    var prependNALUs [][]byte

    for _, nalu := range au {
        typ := h264.NALUType(nalu[0] & 0x1F)

        switch typ {
        case h264.NALUTypeSPS, h264.NALUTypePPS, h264.NALUTypeAccessUnitDelimiter:
            continue
        case h264.NALUTypeIDR:
            // If keyframe, prepend SPS and PPS if available.
            if t.format.SPS != nil && t.format.PPS != nil {
                prependNALUs = [][]byte{t.format.SPS, t.format.PPS}
            }
        }

        // Add other types directly to the filtered NALUs.
        filteredNALUs = append(filteredNALUs, nalu)
    }

    // Prepend SPS and PPS for keyframes at the beginning of the slice.
    if len(prependNALUs) > 0 {
        filteredNALUs = append(prependNALUs, filteredNALUs...)
    }

    return filteredNALUs
}
```

Reply:
I'm not an expert on this code. As I mentioned at the top of the file, this code was taken from mediamtx & is what is used to convert an RTSP stream into something a webrtc connection can consume (which is the goal of the passthrough feature of this viam module).

I'd recommend we get this version of the h264 formatprocessor merged first (as it appears to perform well in practice) and we can optimize / consolidate the code if we find it has any problems.

Reply:
Perhaps a little ambitious to refactor here given that it is based on the mediamtx ref implementation.